### PR TITLE
multi level folders

### PR DESCRIPTION
**Pull request by mdzieg**
added support for folders that are expressed as multiple levels; refactored to reduce code redundancy

### CODE REVIEW

File: FileShareHelper.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            var rootDir = fileShare.GetRootDirectoryReference();
+
+            // Get a reference to the directory.
+            var folder = rootDir.GetDirectoryReference(folderName);
+            var folderExists = await folder.ExistsAsync().ConfigureAwait(false);
+
+            if (!folderExists)
+            {
+                var levels = folderName.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+                var currentLevel = string.Empty;
```
Comment (lucianaparaschivei) [inline FileShareHelper.cs]: I think that you can directly initialise the currentLevel with levels[0] - the root folder

File: FileShareHelper.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+                for (int i = 0; i < levels.Length; i++)
+                {
+                    if (currentLevel == string.Empty) currentLevel = levels[i];
```
Comment (lucianaparaschivei) [inline FileShareHelper.cs]: no need for this line if currentLevel is initialised

**Review by lucianaparaschivei — CHANGES_REQUESTED** [review]
Hello! The refactor looks good 👍 , I only have some minor remarks. Could you also please edit the readme.txt to specify that support for multiple folder levels is added. Thanks.

**Comment by mdzieg** [comment]
Thank you. Will do fixes tomorrow. Will try to also finish off redundancy refactor.

Have to also rethink that fileshare reference passed as a parameter into helper method. Not the cleanest approach but was in a hurry.

Thank you for prompt reaction! Appreciated.

File: FileShareHelper.cs (hunk 1 above)
Comment (mdzieg) [reply FileShareHelper.cs]: I changed the loop to start with the first level. It feels unsafe but the condition above that I added which checks for null or empty and assumption that root level exists 100% should be fine.

File: FileShareHelper.cs (hunk 2 above)
Comment (mdzieg) [reply FileShareHelper.cs]: now I just check the i variable

**Comment by mdzieg** [comment]
readme updated as well
code further refactored to reduce redundancy

File: FileShareManager.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            return folder;
+        }
+        
+        /// <summary>
+        ///     Validates the connection string information in app.config and throws an exception if it looks like
+        ///     the user hasn't updated this to valid values.
+        /// </summary>
+        /// <param name="storageConnectionString">The storage connection string</param>
+        /// <returns>CloudStorageAccount object</returns>
+        private async Task<CloudFileShare> CreateStorageAccountFromConnectionStringAsync(IAzureFileSystemTarget target)
```
Comment (lucianaparaschivei) [inline FileShareManager.cs]: method can be made static

File: FileShareManager.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            }
+
+            return folder;
+        }
+        
+        /// <summary>
+        ///     Validates the connection string information in app.config and throws an exception if it looks like
+        ///     the user hasn't updated this to valid values.
+        /// </summary>
+        /// <param name="storageConnectionString">The storage connection string</param>
```
Comment (lucianaparaschivei) [inline FileShareManager.cs]: adjust the parameter doc to the new parameter

File: FileShareManager.cs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NLog.Extensions.AzureFileSystem
+{
+    public class FileShareManager
+    {
+        private readonly ReaderWriterLockSlim _fileLock = new ReaderWriterLockSlim();
+        private readonly IAzureFileSystemTarget target;
```
Comment (lucianaparaschivei) [inline FileShareManager.cs]: should use _target as variable name

File: AzureFileSystemTarget.cs [hunk 6]
```diff
@@ -16,10 +17,14 @@
 namespace NLog.Extensions.AzureFileSystem
 {
     [Target("AzureFileSystem")]
-    public class AzureFileSystemTarget : TargetWithLayout
+    public class AzureFileSystemTarget : TargetWithLayout, IAzureFileSystemTarget
     {
-        private readonly ReaderWriterLockSlim _fileLock = new ReaderWriterLockSlim();
-        private CloudFileShare _fileShare;
+        private FileShareManager _fileShareManager;
```
Comment (lucianaparaschivei) [inline AzureFileSystemTarget.cs]: can be made readonly

File: AzureFileSystemAsyncTaskTarget.cs [hunk 7]
```diff
@@ -17,13 +17,13 @@
 namespace NLog.Extensions.AzureFileSystem
 {
     [Target("AzureFileSystemAsync")]
-    public class AzureFileSystemAsyncTaskTarget : AsyncTaskTarget
+    public class AzureFileSystemAsyncTaskTarget : AsyncTaskTarget, IAzureFileSystemTarget
     {
-        private readonly ReaderWriterLockSlim _fileLock = new ReaderWriterLockSlim();
-        private CloudFileShare _fileShare;
+        private FileShareManager _fileShareManager;
```
Comment (lucianaparaschivei) [inline AzureFileSystemAsyncTaskTarget.cs]: can be made readonly

**Review by lucianaparaschivei — CHANGES_REQUESTED** [review]
looks good, some small things to address before merging. 👍

File: FileShareManager.cs (hunk 5 above)
Comment (mdzieg) [reply FileShareManager.cs]: fixed

File: FileShareManager.cs (hunk 3 above)
Comment (mdzieg) [reply FileShareManager.cs]: I noticed that this should be parameterless; no point in passing member object. fixed method code

File: AzureFileSystemTarget.cs (hunk 6 above)
Comment (mdzieg) [reply AzureFileSystemTarget.cs]: fixed

File: AzureFileSystemAsyncTaskTarget.cs (hunk 7 above)
Comment (mdzieg) [reply AzureFileSystemAsyncTaskTarget.cs]: fixed

**Pull request merged**